polygon/bridge/bridge_test.go

```diff
@@ -0,0 +1,268 @@
+package bridge_test
+
+import (
+	"context"
+	"errors"
+	"math/big"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/mock/gomock"
+
+	"github.com/ledgerwatch/erigon-lib/common/hexutil"
+	"github.com/ledgerwatch/erigon-lib/log/v3"
+	"github.com/ledgerwatch/erigon/accounts/abi"
+	"github.com/ledgerwatch/erigon/core/types"
+	"github.com/ledgerwatch/erigon/polygon/bor"
+	"github.com/ledgerwatch/erigon/polygon/bor/borcfg"
+	"github.com/ledgerwatch/erigon/polygon/bridge"
+	"github.com/ledgerwatch/erigon/polygon/heimdall"
+	"github.com/ledgerwatch/erigon/polygon/polygoncommon"
+	"github.com/ledgerwatch/erigon/rlp"
+	"github.com/ledgerwatch/erigon/turbo/testlog"
+)
+
+func setup(t *testing.T, abi abi.ABI) (*heimdall.MockHeimdallClient, *bridge.Bridge) {
+	ctrl := gomock.NewController(t)
+	logger := testlog.Logger(t, log.LvlDebug)
+	borConfig := borcfg.BorConfig{
+		Sprint:                map[string]uint64{"0": 2},
+		StateReceiverContract: "0x0000000000000000000000000000000000001001",
+	}
+
+	heimdallClient := heimdall.NewMockHeimdallClient(ctrl)
+	polygonBridgeDB := polygoncommon.NewDatabase(t.TempDir(), logger)
+	store := bridge.NewStore(polygonBridgeDB)
+	b := bridge.NewBridge(store, logger, &borConfig, heimdallClient.FetchStateSyncEvents, abi)
+
+	return heimdallClient, b
+}
+
+func getBlocks(t *testing.T, numBlocks int) []*types.Block {
+	// Feed in new blocks
+	rawBlocks := make([]*types.RawBlock, 0, numBlocks)
+
+	for i := 0; i < numBlocks; i++ {
+		rawBlocks = append(rawBlocks, &types.RawBlock{
+			Header: &types.Header{
+				Number: big.NewInt(int64(i)),
+				Time:   uint64(50 * (i + 1)),
+			},
+			Body: &types.RawBody{},
+		})
+	}
+
+	blocks := make([]*types.Block, len(rawBlocks))
+
+	for i, rawBlock := range rawBlocks {
+		b, err := rawBlock.AsBlock()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		blocks[i] = b
+	}
+
+	return blocks
+}
+
+func TestBridge(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
might be a good idea to add a `t.Cleanup(cancel)` here for the case where:
- the test fails due to a failing assert before calling `cancel()` at the end of the test
- then we will have a goroutine leak because the bridge.Run goroutine won't get the cancel signal
- you won't see any impact of this on single test run but it goroutine leaks may affect CI since there we run a lot of tests at once